tools/omnissm/pkg/manager/manager.go

```diff
@@ -57,27 +86,40 @@ func New(config *Config) *Manager {
 	for _, t := range config.ResourceTags {
 		m.resourceTags[t] = struct{}{}
 	}
+	if config.QueueName != "" {
+		m.queue = NewQueue(config.QueueName, config.Config)
+	}
 	return m
 }
 
 func (m *Manager) Register(doc *identity.Document) (*store.RegistrationEntry, error) {
-	resp, err := m.SSMAPI.CreateActivation(&ssm.CreateActivationInput{
-		DefaultInstanceName: aws.String(doc.Name()),
-		IamRole:             aws.String(m.ssmInstanceRole),
-		Description:         aws.String(doc.Name()),
+	var entry *store.RegistrationEntry
+	err := retry(m.maxRetries, func() error {
+		resp, err := m.SSMAPI.CreateActivation(&ssm.CreateActivationInput{
+			DefaultInstanceName: aws.String(doc.Name()),
+			IamRole:             aws.String(m.ssmInstanceRole),
+			Description:         aws.String(doc.Name()),
+		})
+		if err != nil {
+			return err
+		}
+		entry = &store.RegistrationEntry{
+			Id:             identity.Hash(doc.Name()),
+			ActivationId:   *resp.ActivationId,
+			ActivationCode: *resp.ActivationCode,
+		}
+		return m.Put(entry)
 	})
-	if err != nil {
-		return nil, err
-	}
-	entry := &store.RegistrationEntry{
-		Id:             identity.Hash(doc.Name()),
-		ActivationId:   *resp.ActivationId,
-		ActivationCode: *resp.ActivationCode,
-	}
-	if err := m.Put(entry); err != nil {
-		return nil, err
+	if err == ErrMaxRetriesExceeded {
```

Comment:
this block needs an err = nil before the return out of block on successful queue. 

[edit] it looks like further down its expected to return on this error. semantically that seems odd on successful error handle.

Reply:
so going through the callers, its treated as an error. I think we need a path here that distinguishes an actual unhandled error, versus a deferral case which and ideally a log output to that effect on the later.  The most common case for that is just don't return the error if its handled.

Reply:
if the error isn't returned, for the case it is `ErrMaxRetriesExceeded` when the api call is successfully marshaled and sent to the SQS queue, then the caller for these functions loses the ability to distinguish between successes and failures of the call.  the control flow should be consistent whether the call is deferred to a queue or not (the SQS queue send is optionally applied should the queue be provided/valid).

Reply:
its not boolean though, its really a three way state that the caller has to identify. success, failed, or deferred success, grouping failed and deferred success isn't quite right, the caller is currently treating successful deferral / handled error the same as it would any other error like a network error.  I'd suggest returning a different value nil, nil or a different return value. ie. say you wanted to pass the queue message id back to the client, flowing it on via error is quite odd and assumes differential calling of error all the way up the stack.

Reply:
It seems like it comes down to whether we want "request has been throttled" and "request has been throttled, but you are queued for later action" to be handled differently on the client side. My original assumption was that the client action would be no different (reattempt after some delay), but if we want to change that then I agree the cases would need to be differentiated. From the client's perspective this case is still distinct from a network error, unless I'm missing something: in the case of a network issue the stdlib http calls return an error, whereas if we return an error from our API the client sees a "successful" request that happens to have status code == 500 (and potentially a descriptive response body).

Reply:
There is nothing odd about returning the error here.  It would be also very bad to set the err to nil on successful queue because, the caller can no longer distinguish 100% what happened based upon what was returned.  The right thing to do here is to return the error because the error isn't being handled, rather there is conditional behavior that is happening should an error occur in the function.  The error is being handled when it is being outputted to the logger by the caller.

Reply:
a 500 status code to the client is also impossible to distinguish between perm error and retry (ala 503 or otherwise), say for example someone does ops paging on status code. Logging is not handling an error, there's 20 different errors that may pop in this code, that are actual errors, successful queue is not an error, its a normal path. golang is multi-return, so distinguishing on the error or the value is semantics, the issue is a common path should return a meaningful differential to the client, rather than treating unexpected internal errors the same as something thats been explicitly handled.